alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,50 @@
+# type: ignore
+from alibi_detect.version import __version__
+from alibi_detect.utils.schemas import DETECTOR_CONFIGS, DETECTOR_CONFIGS_RESOLVED, __config_spec__
+import warnings
+
+
+def validate_config(cfg: dict, resolved: bool = False) -> dict:
+    """
+    Validates a detector config dict by passing the dict to the detector's pydantic model schema.
+
+    Parameters
+    ----------
+    cfg
+        The detector config dict.
+    resolved
+        Whether the config is resolved or not. For example, if resolved=True, `x_ref` is expected to be a
+        np.ndarray, wheras if resolved=False, `x_ref` is expected to be a str.
+
+    Returns
+    -------
+    The validated config dict, with missing fields set to their default values.
+    """
+    # Get detector name
+    if 'name' in cfg:
+        detector_name = cfg['name']
+    else:
+        raise ValueError('`name` missing from config.toml.')
+
+    # Validate detector specific config
+    if detector_name in DETECTOR_CONFIGS.keys():
+        if resolved:
+            cfg = DETECTOR_CONFIGS_RESOLVED[detector_name](**cfg).dict()  # type: ignore[attr-defined]
+        else:
+            cfg = DETECTOR_CONFIGS[detector_name](**cfg).dict()  # type: ignore[attr-defined]
+    else:
+        raise ValueError('Loading the specified detector from a config.toml is not yet supported.')
+
+    # check version
+    version = cfg.pop('version', None)
+    if version is not None and version != __version__:
```

Comment:
Good idea IMO.

Reply:
@mauicv have some thoughts on this now, will get back to you tmrow!

Reply:
@jklaise @mauicv 
I'm now rethinking if we even need `config_spec`. We can't update `config_spec` without updating `version`, so the only valid use case to have both is if we wish to update `version` but not `config_spec`. This would make sense if we wished to gurantee config functionality across `version`'s, but not `config_spec`'s. However, I'm not sure we do want to do that, as save/loading across `version`'s is just as likely to cause problems. 

Another way to put it is, since the config is so tightly coupled with our API, do we actually need both `config_spec` and `version` number?

Reply:
Good question. I thought initially it would make sense as we wouldn't change `config_spec` every version and it would be nice to be able to have compatibility between "adjacent" versions. But would we be shooting ourselves in the foot by doing it? Are there examples where config wouldn't change but the saving/loading could be broken due to some other stuff changing?

A related rather gnarly issue I just thought of is artefacts of dependencies. I.e. we are compatible with many `tensorflow` or `torch` versions, but our config doesn't know which version the artefacts were saved with, so it's possible that the library version is the same but due to a change in dependency version some artefacts are broken. I guess this should align more with the "version your environment" philosophy though, not saying necessarily we should track the versions of the dependency artefacts as you can take it arbitrarily far (e.g. saving exact Python version too).

Reply:
Maybe it makes sense to only have the `config_spec` and have `version` only relate to the `api`. If a version update to the `api` also corresponds to breaking changes to the save load functionality then we update the `config_spec`? We're only really looking to track breaking changes right? So we bump the `config_spec` when we want to notify users they have to update there config files to work with later versions of the `api`? Although I guess those might just be considered Major changes? hmmm 🤔

Reply:
Based on offline discussion, we will retain `config_spec`, but only major-minor version numbers. Documentation of the intended use has been added to `alibi_detect/version.py`.